File: scripts/settle.py
#!python3
import math
from web3 import Web3
from dotenv import load_dotenv
from datetime import datetime
import json
import os
import requests

load_dotenv()

node = os.getenv('GOERLI_URL')
web3 = Web3(Web3.HTTPProvider(node))

# https://thegraph.com/hosted-service/subgraph/horse-link/hl-protocol-goerli
bets_query = """
{
  bets(where: {createdAt_gt: %(createdAt_gt)s, settled: false},orderBy:createdAt) {
    id
    createdAt
    createdAtTx
    marketId
    marketAddress
  }
}
"""


def get_subgraph_bets_since(createdAt_gt):
    query = bets_query % {"createdAt_gt": createdAt_gt}
    response = requests.post(
        "https://api.thegraph.com/subgraphs/name/horse-link/hl-protocol-goerli",
        data=json.dumps({"query": query}),
    )
    data = json.loads(response.text)["data"]
    return data["bets"]


def hydrate_market_id(market_id):
    # Remove the '0x' prefix
    market_id = market_id[2:]

    # Convert hexadecimal to binary
    binary = bytes.fromhex(market_id)
    # Decode binary as ASCII
    market_string = binary.decode("ascii")
    print(f">{market_string}<")

    # Parse the market string
    id = market_string[0:11]
    date = int(market_string[0:6])
    location = market_string[6:9]
    race = int(market_string[9:11])
    # Return a dictionary with the hydrated market data
    result = {
        "id": str(id),
        "date": date,
        "location": location,
        "race": race
    }
    return result


def get_oracle():
    response = requests.get('https://alpha.horse.link/api/config')
    data = response.json()
    return data['addresses']['marketOracle']


def load_market(address, token):
    response = requests.get(
        f'https://raw.githubusercontent.com/horse-link/contracts.horse.link/main/deployments/goerli/{token}Market.json')
    data = response.json()
    abi = data['abi']
    contract = web3.eth.contract(address=address, abi=abi)
    return contract


def load_oracle():

    address = get_oracle()

    response = requests.get(
        'https://raw.githubusercontent.com/horse-link/contracts.horse.link/main/deployments/goerli/MarketOracle.json')
    data = response.json()
    abi = data['abi']
    contract = web3.eth.contract(address=address, abi=abi)
    return contract


def get_result(oracle, marketId):
    # id as byte array
    encoded = marketId.encode('utf-8')
    id = bytearray(encoded)

    result = oracle.functions.getResult(id).call()
    return result


def set_result(oracle, marketId, propositionId, signature) -> None:
    try:
        # Can be any account with funds
        account_from = {
            'private_key': os.getenv('SETTLE_PRIVATE_KEY'),
            'address': '0xF33b9A4efA380Df3B435f755DD2C2AF7fE53C2d1',
        }

        signature_tuple = [signature['v'], signature['r'], signature['s']]

        encoded = marketId.encode('utf-8')
        id = bytearray(encoded)

        encoded = propositionId.encode('utf-8')
        proposition_id = bytearray(encoded)

        tx = oracle.functions.setResult(id, proposition_id, signature_tuple).buildTransaction(
            {
                'from': account_from['address'],
                'nonce': web3.eth.get_transaction_count(account_from['address']),
            }
        )

        tx_create = web3.eth.account.sign_transaction(
            tx, account_from['private_key'])

        tx_hash = web3.eth.send_raw_transaction(tx_create.rawTransaction)
        tx_receipt = web3.eth.wait_for_transaction_receipt(tx_hash)

        print(tx_receipt)
    except:
        print("An exception occurred")


def settle(market, index):
    # Can be any account with funds
    account_from = {
        'private_key': os.getenv('SETTLE_PRIVATE_KEY'),
        'address': '0xF33b9A4efA380Df3B435f755DD2C2AF7fE53C2d1',
    }

    tx = market.functions.settle(index).buildTransaction(
        {
            'from': account_from['address'],
            'nonce': web3.eth.get_transaction_count(account_from['address']),
        }
    )

    tx_create = web3.eth.account.sign_transaction(
        tx, account_from['private_key'])

    tx_hash = web3.eth.send_raw_transaction(tx_create.rawTransaction)
    tx_receipt = web3.eth.wait_for_transaction_receipt(tx_hash)

    return tx_receipt


def main():
    # fetch registry contract address from the api
    oracle = load_oracle()
    now = datetime.now().timestamp()
    print(f"Current Time: {now}")

    # Now less 2 hours
    close_time = math.floor(now) - 7200
    print(f"Using close time of {close_time}")

    bets = get_subgraph_bets_since(close_time)

    for bet in bets:
        # check if bet is settled via the api
        # hydrate
        hydrated_market = hydrate_market_id(bet['marketId'])
        id = hydrated_market["id"]

        # call api to get result
        response = requests.get(f'{id}')

        # check if result has been added to the oracle
        result = get_result(oracle, id)

        # If we have a result from the API and the oracle has not already added the result
        if response.status_code == 200 and result != b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00':
            print(f"Settling bet {bet[id]} for market {bet['marketAddress']}")

            tx_receipt = settle(market, i)
            print(tx_receipt)


if __name__ == '__main__':
    print('Starting settle script')
    main()

File: scripts/settle.ts
import dotenv from "dotenv";
import fs from "fs";
import {
	getSubgraphBetsSince,
	loadOracle,
	hydrateMarketId,
	loadMarket,
	Seconds,
	bytes16HexToString,
	setProvider,
	setAxiosClient,
	axiosClient
} from "./utils";
import type { MarketDetails } from "./utils";
import type { AxiosResponse } from "axios";

dotenv.config();
const hexZero: Bytes16 = "0x00000000000000000000000000000000";
const HOUR_IN_SECONDS = 60 * 60;

export type RaceDetails = {
	id: string;
	market: MarketDetails;
	number: string;
};

export type BetDetails = {
	id: string;
	createdAt: string;
	createdAtTx: string;
	marketId: string;
	marketAddress: string;
};

export type Milliseconds = number;
export type Bytes16 = string;

// TODO: Add a better type for this data structure:
// [
// 	'0x00000000000000000000000000000000',
// 	[],
// 	winningPropositionId: '0x00000000000000000000000000000000',
// 	scratched: []
//   ]
// that's result[0],result[1], result.winningPropositionId, result.scratched
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type OracleResult = Array<any>;

export async function main() {
	const deploymentName = process.argv[2];
	const { chainId, baseApiUrl, subgraphUrl, privateKeyEnvVar, providerUrl } =
		JSON.parse(fs.readFileSync(`./config_${deploymentName}.json`).toString());
	setProvider(providerUrl);
	setAxiosClient(chainId, baseApiUrl);

	const oracle = await loadOracle(deploymentName, privateKeyEnvVar);
	const now: Seconds = Math.floor(Date.now() / 1000);
	console.log(`Current Time: ${now} (seconds)`);

	const closeTime: Seconds = now - 48 * HOUR_IN_SECONDS;
	console.log(`"Using close time of ${closeTime} (seconds)"`);

	const bets: BetDetails[] = await getSubgraphBetsSince(
		subgraphUrl,
		closeTime,
		{
			unsettledOnly: true,
			maxResults: 150,
			payoutAtLt: now
		}
	);

	console.log(`Found ${bets.length} unsettled bets`);

	for (const bet of bets) {
		const market = hydrateMarketId(bet.marketId);
		// TODO: cache me
		const marketContract = await loadMarket(
			deploymentName,
			bet.marketAddress,
			privateKeyEnvVar
		);

		let marketResultResponse: AxiosResponse;
		// Get race result
		// TODO: Only query each market once
		try {
			marketResultResponse = await axiosClient.get(
				`/markets/result/${market.id}?sign=true`
			);
			if (marketResultResponse.status !== 200) {
				console.log(`request failure for market ${market.id}:`, {
					status: marketResultResponse?.status,
					statusText: marketResultResponse?.statusText,
					url: marketResultResponse?.config?.url,
					data: marketResultResponse?.request?.data
				});
				continue;
			}
			// eslint-disable-next-line @typescript-eslint/no-explicit-any
		} catch (e: any) {
			// TODO: type as axios.AxiosError;
			console.log(`request failure for market ${market.id}:`, {
				responseStatus: e?.response?.status,
				responseStatusText: e?.response?.statusText,
				url: e?.config?.url,
				data: e?.request?.data
			});
			continue;
		}

		let result;
		try {
			result = await oracle.getResult(bet.marketId);
		} catch (e) {
			console.log("getResult failed:", JSON.stringify(e));
			continue;
		}

		if (result === undefined) {
			console.log(
				`no result from oracle.getResult() for marketId ${
					bet.marketId
				}(${bytes16HexToString(bet.marketId)})`
			);
			continue;
		}

		if (result.winningPropositionId === hexZero) {
			// if the oracle doesn't know about it, tell it about it.
			try {
				const txReceipt = await oracle.setResult(
					bet.marketId,
					marketResultResponse.data.winningPropositionId,
					marketResultResponse.data.signature
				);
				console.log(`adding result for market ${bet.marketId}`, txReceipt.hash);
			} catch (e) {
				console.log("setResult failed:", JSON.stringify(e));
				continue;
			}

			// we won't wait for it to be mined here,
			// we'll just process it on the next run
		} else {
			// There is a result so we can settle

			const index = bet.id.split("_")[2];
			const txReceipt = await marketContract.settle(index).catch((e) => {
				return { hash: `FAILED: ${e?.error?.reason}` };
			});
			console.log(`settle bet ${bet.id}(${index}), receipt`, txReceipt.hash);
		}
	}
}

// This is a translation of a pythonism: `ts-node settle.ts` will run main(),
// but import/require("settle") won't
if (require.main === module) {
	main()
		.then(() => {
			console.log("Done");
			process.exit(0);
		})
		.catch((error) => {
			console.error(error);
			process.exit(1);
		});
}
